NeoML/src/Dnn/Layers/TiedEmbeddingsLayer.cpp

```diff
@@ -0,0 +1,162 @@
+/* Copyright � 2017-2020 ABBYY Production LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#include <common.h>
+#pragma hdrstop
+
+#include <NeoML/Dnn/Layers/TiedEmbeddingsLayer.h>
+#include <NeoML/Dnn/Layers/MultichannelLookupLayer.h>
+
+namespace NeoML {
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+CTiedEmbeddingsLayer::CTiedEmbeddingsLayer( IMathEngine& mathEngine ) :
+	CBaseLayer( mathEngine, "CTiedEmbeddingsLayer", true ),
+	channelIndex( 0 )
+{
+}
+
+void CTiedEmbeddingsLayer::SetChannelIndex( int val )
+{
+	NeoAssert( val >= 0 );
+
+	channelIndex = val;
+}
+
+static const int CnnTiedEmbeddingsLayerVersion = 2000;
+
+void CTiedEmbeddingsLayer::Serialize( CArchive& archive )
+{
+	archive.SerializeVersion( CnnTiedEmbeddingsLayerVersion, CDnn::ArchiveMinSupportedVersion );
+	CBaseLayer::Serialize( archive );
+
+	archive.Serialize( embeddingsLayerName );
+	archive.Serialize( channelIndex );
+}
+
+void CTiedEmbeddingsLayer::Reshape()
+{
+	CheckInputs();
+
+	CheckArchitecture( GetDnn()->HasLayer( embeddingsLayerName ), GetName(),
+		"Wrong name for embeddings layer." );
+
+	const int embeddingsChannelsCount = CheckCast<CMultichannelLookupLayer>(
```

Comment:
It's better to do `embeddingsLayer = dynamic_cast<CMultichannelLookupLayer*>` and `CheckArchitecture( embeddingsLayer != nullptr )`. In that case the error will be more informative: wrong layer type instead of `somewhere CheckCast has failed`. Especially useful in version without FineObj.